### PR TITLE
Curved Labels

### PR DESCRIPTION
**Pull request by dmvaldman**
This PR introduces curved label support for Tangram and reverts the previous "articulated" labels support, where labels could be kinked on a single space. @bcamper @nvkelso

www gifcreator me_gu1kbd

**Architecture**

Text from a road label is split into 2 or 3 letter pairs. Each of which is rendered as its own quad along the road geometry. These quads need extra data to render than what we've traditionally used. As a user zooms the map, the angles and offsets of these quads needs to change dynamically, which isn't true for other quads such as those using our points style or non-curved labels. 

Since our current architecture doesn't allow us to modify GL buffers as the user is zooming, we need to precompute any relevant data from the un-zoomed tile that we may need at a fractional zoom. We calculate the exact angle and offset of each quad relative to the un-zoomed tile at three fractional zoom levels: .33, .66 and .99. 

|zoom| image|
|--|--|
|17.0||
|17.33||
|17.66||
|17.99||

If this was all we did, the labels would "snap" into place at these discrete zoom levels, and look badly between the zoom levels. To remedy this, we interpolate the values of the offsets and angles in the vertex shader, so that they vary continuously between the zoom steps. This can still lead to interpolation artifacts in some cases where for very windy roads, the interpolation is not accurate enough. 

I will forever be indebted to @patriciogonzalezvivo for this [shader interpolation technique](https://github.com/tangrams/tangram/blob/labels-rebase/src/styles/points/points_vertex.glsl#L70-L77).

**Language Support**

Not all languages can make use of our curved label architecture. Languages with context-shaping (where characters have different shapes depending on their neighboring characters) have to be excluded. This excludes: Arabic, Bengali, Burmese, Khmer, Lao, Mongolian, Tibetan and the following Indian dialects: Devangari, Gujarati, Gurmukhi, Kannada, Oriya, Tamil, Telugu (@tallytalwar am I missed anything here?). A handy reference for the various script features of languages can be found [here](http://r12a.github.io/scripts/featurelist/).

Labels with mixed RTL and LTR text are also excluded, however labels of entirely RTL/neutral or LTR/neutral text are handled. Notably, Hebrew, a RTL language without context-shaping, is handled appropriately.

**Finding a "Good" Curved Label**

To select a label, first we try to find a straight label that is no larger than 50% ([variable](https://github.com/tangrams/tangram/blob/labels-rebase/src/labels/label_line.js#L6)) of the longest line segment in the road's line geometry (Note: this is subtlety different than before, where a label would have to fit within some percentage of a single segment. We can now group multiple segments together if they are co-linear). If no such label is found, we look for the "straightest" part of the line geometry. To do this, we build a [cost function](https://github.com/tangrams/tangram/blob/labels-rebase/src/labels/label_line.js#L376) that measures the amount of curvature. For each vertex of the line (where a curved label with start), we create a window that is the width of the label, and assign a weight to it that is the sum of the curvature of each segment falling within the window. We then select the window with the smallest cost, that also satisfies a maximum curvature at any particular segment, and place the label there. This yields good results, but finding the right parameters can be something of an art.

There is still the small probability that two characters can overlap on a label. I have yet to see this in my debugging though because I am keeping the fit tolerances described above tight. Ultimately the probability of this is font-dependent.

**Stroking Curved Labels**

For curved labels with stroking, we need to render them in two passes: first the strokes, then the fills. Otherwise, the stroke of one label fragment will overlap on the filled text of the previous label fragment.

For example here is a tile and its corresponding texture atlas

### CODE REVIEW

**Comment by patriciogonzalezvivo** [comment]
EPIC PR!!!

**Comment by bcamper** [comment]
@nvkelso can you help review the placement differences here? From first glance, it seems like we are gaining in some areas (yay), but losing in others, and we may need to tweak internal thresholds to get the right balance. Some of these look too sparse (from my quick look):

http://tangrams.github.io/differ/?1=https://tangrams.github.io/bubble-wrap/bubble-wrap-style.yaml&2=&lib1=0.11&lib2=https://precog.mapzen.com/tangrams/tangram/labels-rebase/dist/tangram.debug.js&ignore1&ignore2&go

**Comment by bcamper** [comment]
@dmvaldman did we end up with any performance numbers to compare here?

**Comment by dmvaldman** [comment]
Will get you some of those. Any things in particular you'd like me to test?

**Comment by nvkelso** [comment]
Will review and provide back comments.

**Comment by nvkelso** [comment]
(I'll try and add some before/after images, but wanted to get you this list of test locations now.)

Suspected not enough overlap allowance:

1. Vinton Ct. https://tangrams.github.io/bubble-wrap/#17/37.79144/-122.40247
  But other small streets with a little overlap do show up, so I'm confused.
2. Alton Ave. https://tangrams.github.io/bubble-wrap/#15/37.7561/-122.4512
3. 19th Ave. https://tangrams.github.io/bubble-wrap/#15/37.7553/-122.4546
  This is the main north-south highway in western San Francisco
4. 3rd St: https://tangrams.github.io/bubble-wrap/#11/37.7715/-122.3547
  But overall zoom 11 is hugely better as-is!

Complete disappearance (maybe overlap related?):

1. Starview Wy., Knollview Wy., Gladeview Wy: https://tangrams.github.io/bubble-wrap/#17/37.75104/-122.44879
  Zooming in on this area does reveal the labels, but it seems like their spacing is stuck at z16 or z18 even when zoomed into zoom 20 (they are too far apart)
2. Twin Peaks Blvd: https://tangrams.github.io/bubble-wrap/#15/37.7561/-122.4512
3. Portola and Market: https://tangrams.github.io/bubble-wrap/#14/37.7496/-122.4672
4. Howard, Folsom, Divisadero, 16th: https://tangrams.github.io/bubble-wrap/#14/37.7908/-122.4395
  Major regressions at zoom 14 across the board. 
4. Conzelman Rd. gone: https://tangrams.github.io/bubble-wrap/#15/37.8290/-122.5152
  And less than expected at zoom 16: https://tangrams.github.io/bubble-wrap/#16/37.8301/-122.5097
  Zoom 17 is good
5. Ocean, Tunnel, Central: https://tangrams.github.io/bubble-wrap/#12/37.7746/-122.4371
  Minor smattering of roads that could use more overlap?
5. Sausalito z15: https://tangrams.github.io/bubble-wrap/#15/37.8542/-122.4851
5. Sausalito z16: https://tangrams.github.io/bubble-wrap/#16/37.8530/-122.4852
6. Sausalito z17: https://tangrams.github.io/bubble-wrap/#17/37.85687/-122.49126
7. More than 50% reduction in labels in Mill Valley at zoom 15! https://tangrams.github.io/bubble-wrap/#15/37.9052/-122.5496
8. Mill Valley z16: https://tangrams.github.io/bubble-wrap/#16/37.9066/-122.5555
  FWIW: Apple Maps also has problems with all the curvy roads here, and a Google Maps is only slightly better than Apple in Mill Valley street labels.
9. Many labels missing from London: https://tangrams.github.io/bubble-wrap/#13/51.5067/-0.1004
10. London z14: https://tangrams.github.io/bubble-wrap/#14/51.5119/-0.1292
11. London z15: https://tangrams.github.io/bubble-wrap/#15/51.5146/-0.1300
  This zooms seems to remove some, add some, but some big streets are missing
12. London z16: https://tangrams.github.io/bubble-wrap/#16/51.5154/-0.1356
  Similar to z15.
13. London z17: https://tangrams.github.io/bubble-wrap/#17/51.51376/-0.14339
  Surprising that this continues to be substantially different than before.

Some street labels added, good!

1. Sunset, Noriega, 9th: https://tangrams.github.io/bubble-wrap/#14/37.7496/-122.4672
  But same zoom in London shows significant regression :\  
2. Everywhere in SF: https://tangrams.github.io/bubble-wrap/#13/37.7747/-122.4373
  A huge improvement! The long highway names that are "missing" here are a good improvement, too.
  But same zoom in London shows significant regression :\

**Comment by nvkelso** [comment]
(In some cases letter collide each other on curves, I'll add some specific examples.)

**Comment by dmvaldman** [comment]
Thanks for the thoroughness @nvkelso. Though some of these regressions are a result of stricter label overflow settings (perhaps this is what you mean by "overlap"?) and the precariousness of our collision pass, some others are may indicate a bug. I'll get back to you with an update once I investigate things more closely.

**Comment by bcamper** [comment]
Thanks @dmvaldman -- what I think would work (which I think is related to what you're saying above) is to allow a relaxed tolerance for straight labels, but ranked lower than some curved labels. So a ranking like:

1. Ideal straight label
2. Curved label
3. Tolerable straight label (more overflow, but OK when neither of the above two fits)

**Comment by dmvaldman** [comment]
Voilà https://precog.mapzen.com/tangrams/tangram/labels-rebase-loose-overlap

**Comment by nvkelso** [comment]
_Having a look..._

**Comment by nvkelso** [comment]
Looking much better!

Other areas for improvement, below? 

1. Market St. and Woodside Ave missing? https://tangrams.github.io/bubble-wrap/#15/37.7553/-122.4546
  screen shot 2017-02-02 at 13 30 15
2. Twin Peaks street label at z16 on curvy bit: https://tangrams.github.io/bubble-wrap/#16/37.7524/-122.4475
  Apple doesn't have, Google does. Seems not so curvy.
  screen shot 2017-02-02 at 13 39 32
3. Conzelman Rd. is back! But still not many placements, and some nasty character collisions: https://tangrams.github.io/bubble-wrap/#15/37.8290/-122.5152
  screen shot 2017-02-02 at 13 45 45
4. These all still missing: Ocean, Tunnel, Central: https://tangrams.github.io/bubble-wrap/#12/37.7746/-122.4371
  But overall other labels added that are better than before so, eh? 
5. Would be nice to see articulated Sausalito Blvd z15: https://tangrams.github.io/bubble-wrap/#15/37.8542/-122.4851
  screen shot 2017-02-02 at 13 49 12
6. z16 in Saucalito looks hugely better (including adding the articulated Sausalito Blvd label), though there several collided characters (16/37.8543/-122.4865)
  screen shot 2017-02-02 at 13 50 27
  screen shot 2017-02-02 at 13 50 23
7. Mill Valley z15 is better, but still missing important streets like Panoramic Highway and Edgewood https://tangrams.github.io/bubble-wrap/#15/37.9052/-122.5496
  screen shot 2017-02-02 at 13 55 21
8. Still too few labels in Mill Valley z16: https://tangrams.github.io/bubble-wrap/#16/37.9066/-122.5555
  Maybe if road is very curvy allow it to overlap more? Kink more?
9. Still a problem in London with many labels missing z13: https://tangrams.github.io/bubble-wrap/#13/51.5067/-0.1004
10. London z14: https://tangrams.github.io/bubble-wrap/#14/51.5119/-0.1292
  Shows some improvement, could still be a little better.

_I'll followup with more verification, additional soon..._

**Comment by nvkelso** [comment]
Looks like we're dropping many labels in Tokyo? https://tangrams.github.io/bubble-wrap/#12/35.6733/139.7131

**Comment by dmvaldman** [comment]
I thought there may have been a bug with the Tokyo example, but when I used the latest bubblewrap style, things look comparable in label density to before. I think it may be a scene-file specific thing you saw of now showing more minor roads at higher zooms. @nvkelso 

**Current**

tangram-12-35 67333-139 71313-20170271852

**Curved Label Branch**

tangram-1486511573778
 
I've also loosened the overlap allowance for the second pass on straight labels. We should have even more straight labels appearing now.

**Comment by nvkelso** [comment]
**tl:dr:** 

- an overlap value of `2.3` for the final straight line pass seems to give the best results.

**discussion:**

Using a related branch (precog.mapzen.com/tangrams/tangram/labels-rebase-test/) I've experimented with the overlap values to encourage more hard to place labels to still place.

There are several label passes now in Tangram:

1. Straight labels with an allowed max overlap of `1.5`
2. Curved labels
3. Straight labels pass with an allowed max overlap of `3.0`

methodology: @dmvaldman setup a console option for me to adjust the values of the final straight line pass.

- `scene.config.global.overflow = 2;`
- `scene.updateConfig()`

After looking at the test locations in https://github.com/tangrams/tangram/pull/484#issuecomment-277098509, an overlap value of `2.3` looks best to me (most labels without egregious overlaps).

1. Woodside Ave returns with 2.2 (Market St. still gone until 3.0): https://tangrams.github.io/bubble-wrap/#15/37.7553/-122.4546
2. Twin Peaks street label at z16 on curvy bit only at 3.8 (SKIP): https://tangrams.github.io/bubble-wrap/#16/37.7524/-122.4475
3. Conzelman Rd. at 2 is okay, 2.5 is better, 2.2 is okay (has larger impact at zoom 16): https://tangrams.github.io/bubble-wrap/#15/37.8290/-122.5152
4. These all still missing: Ocean, Tunnel, Central (not added back until 2.7 only for Ocean, other still missing): https://tangrams.github.io/bubble-wrap/#12/37.7746/-122.4371
5. Would be nice to see articulated Sausalito Blvd z15 (2.9 for straight): https://tangrams.github.io/bubble-wrap/#15/37.8542/-122.4851
6. z16 in Saucalito looks hugely better (including adding the articulated Sausalito Blvd label), though there several collided characters (16/37.8543/-122.4865)
7. Mill Valley z15 is better, but still missing important streets like Panoramic Highway (2.1) and Edgewood (3.0) https://tangrams.github.io/bubble-wrap/#15/37.9052/-122.5496
8. Still too few labels in Mill Valley z16 (okay 2.2, better 2.3): https://tangrams.github.io/bubble-wrap/#16/37.9066/-122.5555
9. Still a problem in London with many labels missing z13 (2.3 good): https://tangrams.github.io/bubble-wrap/#13/51.5067/-0.1004
10. London z14 (2.3 good): https://tangrams.github.io/bubble-wrap/#14/51.5119/-0.1292
11. London z15 (2.3 good): https://tangrams.github.io/bubble-wrap/#15/51.5146/-0.1300

**Comment by nvkelso** [comment]
@dmvaldman I wonder if there are actually 4 passes?

There are several label passes now in Tangram:

1. Straight labels with an allowed max overlap of `1.5`
2. Curved labels
3. ~~Articulated labels~~ 
4. Straight labels pass with an allowed max overlap of ~~3.0~~ `2.3`

I think we removed Articulated totally, but I am seeing some areas where it's not replaced with any labels, like Bradley Blvd below – maybe this has it's own cutoff criteria?

http://tangrams.github.io/bubble-wrap/#14/38.9810/-77.1024

screen shot 2017-02-08 at 16 11 47

**Comment by nvkelso** [comment]
Confirmed that `Bradley Blvd` example in https://github.com/tangrams/tangram/pull/484#issuecomment-278506017 above is a red herring and on me to fix in the Mapzen basemap styles and not related to this Tangram PR. Turns out that is a config error in the src YAML (wrong label defaults in src code, correct in JS rig).

**Comment by dmvaldman** [comment]
Thanks @nvkelso! It would be good to know if there are other blockers to the PR being merged in (besides setting the overlap parameter to 2.3). For instance: letter collisions, or other places where labels should be appearing but aren't. I'm happy to expose some letter collision parameters as I did the overlap parameter for testing.

**Comment by nvkelso** [comment]
While I'm happy to test letter collisions with more global options, depends on your release schedule. What you have now (after the `2.3` overlap change) will be a huge improvement and is worth getting out sooner. But if it's not too much trouble and there's another week I'm happy to dial collisions up / down.

The two areas with character collisions noted above were (I've seen others, too):

1. Conzelman Rd. is back! But still not many placements, and some nasty character collisions: https://tangrams.github.io/bubble-wrap/#15/37.8290/-122.5152
  - screen shot 2017-02-02 at 13 45 45
2. z16 in Saucalito looks hugely better (including adding the articulated Sausalito Blvd label), though there several collided characters (16/37.8543/-122.4865)
  - screen shot 2017-02-02 at 13 50 27
  - screen shot 2017-02-02 at 13 50 23

**Comment by bcamper** [comment]
📰

**Comment by meetar** [comment]
📈

**Comment by dmvaldman** [comment]
My PR might be the one that is being merged, but this is the concerted effort of so many. I'd like to thank @hjanetzek for his insightful comments, @patriciogonzalezvivo for his epic shader interpolation, without which none of this would be performant, @tallytalwar for sanity-checking labels in India, @nvkelso for his tireless testing and @bcamper for believing in me and providing me courage on this twisted adventure. There are so many others to thank, oh gosh, thank you Mapzen, my friends and family who have put up with my incessant banter that text "isn't as it appears", the academy, .... *{walks of stage in tears}*

**Comment by louh** [comment]
epic!!

**Comment by bcamper** [comment]
Released in [v0.12.0](https://github.com/tangrams/tangram/releases/tag/v0.12.0)
